### DISCUSSION TITLE
useMatches and Typescript

### DISCUSSION THREAD

**Discussion by bnabet** [remix-run/react-router#9534] [Q&A] [answered]
Hello,

I don't understand the data returned by the library. I'm looking to develop a breadcrumb but I'm having trouble trying to use it with Typescript.

Let's take a **Home** route like this :
```tsx
<Route
  path="/"
  element={<Home />}
  handle={{
    infos: {
      title: "Home page"
    }
  }}
/>
```

And a component **Breadcrumbs** :
```tsx
import { RouteObject, useMatches } from "react-router-dom";

function Breadcrumbs() {
  const matches: RouteObject[] = useMatches();

  const routesInfos = matches
    .filter((match) => Boolean(match.handle?.infos))
    .map((match) => {
      console.log(match);

      return {
        title: match.handle.infos.title,
        path: match.pathname
      };
    });
}
```

Here is the return of the **match** log.
First thing, we can see that `path` prop becomes `pathname`.
```json
{
    "id": "0-0",
    "pathname": "/",
    "params": {},
    "handle": {
        "infos": {
            "title": "Home page"
        }
    }
}
```

But now i get a Typescript error.
![Capture d’écran 2022-11-01 à 13 35 56](https://user-images.githubusercontent.com/14838747/199233866-d6a45989-3939-425b-b03d-fa39b0cf3239.png)

What we can indeed confirm when we look at the types :
```ts
export interface IndexRouteObject {
    caseSensitive?: AgnosticIndexRouteObject["caseSensitive"];
    path?: AgnosticIndexRouteObject["path"];
    id?: AgnosticIndexRouteObject["id"];
    loader?: AgnosticIndexRouteObject["loader"];
    action?: AgnosticIndexRouteObject["action"];
    hasErrorBoundary?: AgnosticIndexRouteObject["hasErrorBoundary"];
    shouldRevalidate?: AgnosticIndexRouteObject["shouldRevalidate"];
    handle?: AgnosticIndexRouteObject["handle"];
    index: true;
    children?: undefined;
    element?: React.ReactNode | null;
    errorElement?: React.ReactNode | null;
}
```

So I find myself stuck.
I have to define the matches as `RouteObject` but the matches returns me `pathname` while it expects a `path`.

Sorry if my explanation is not very clear and thank you in advance for your help.

Have a nice day !

**Comment by drenther** [comment]
Facing issues typing this as well?

Is there a good workaround for this?

**Comment by bnabet** [comment]
Thank you for your comment.

I have not found any other solution than to completely rethink my component. I think the basis of my problem is a wrong imported type. But the documentation for Typescript is so poor...

I'm still looking for the correct solution.

**Comment by penx** [comment]
Try something like this:


```ts
type RouteWithHandle<Handle extends string, Value> = {
  id: string;
  pathname: string;
  params: Params<string>;
  data: unknown;
  handle: Record<Handle, Value>;
};

function isRecordWithKey<T extends string>(
  value: unknown,
  key: T,
): value is Record<T, unknown> {
  return !!value && typeof value === "object" && key in value;
}

function hasHandle<Handle extends string, Value>(
  handle: Handle,
  valuePredicate?: (v: unknown) => v is Value,
) {
  return (
    route:
      | {
          handle: unknown;
        }
      | RouteWithHandle<Handle, Value>,
  ): route is RouteWithHandle<Handle, Value> => {
    return (
      !!route.handle &&
      isRecordWithKey(route.handle, handle) &&
      (!valuePredicate ||
        (handle in route.handle && valuePredicate(route.handle[handle])))
    );
  };
}

function isString(value: unknown): value is string {
  return typeof value === 'string';
}

function MyComponent() {
  const matches = useMatches();
  const routesInfos = matches.filter(hasHandle("infos", isString)).map(match => match.handle.infos);
}
```

[playground](https://www.typescriptlang.org/play?#code/PQKhAIBcE8AcFMDO4BmAnA9gW3G+BDAY0gFpMBXSeNAAQDYA6AFnDGACh4APWDNScABN4hADb48UOPHAAFCfiyIAPAGl40cNyoA7QckSQ0ASx0BzcAF5whk+YB8V8AG924d7gKCMO0ZoDaANYa4Kbg6tAAugBcNkamFgA+4OR68Cim8IIA3OwAvrmEPoYpiPAAsviQhAAWSE4AFACUVo4N-pHg+MiuHqGCsbYJuX2wVTU6ivCD8eYjHmNoioix8ktKykMO8+6CVfixqYE6GADuOjvgNfh6otMpOsdnF-kdLeygECTfJKwgHOwYAhwAAlDCUeAAdWMkBqAAkboI7soEbcZNp4HoDLMzAAacAANXwonI8Ec1l6-RmdjM8zGsMmWHuW1pbnAi2WqwUGxZ9nme0gBweT3O82uaNiIJEfEEKMRd3xRJJZNyBXY7BQqWIxh8oUQUqKaEE0NhEWUABUtFxdPo4jT7A02QA3Ymkw6PE7nXFs4LQWLm71NWIu5V60HSo0W-FHT06RyUvCQchoHTgACEaZDpPAADIc1IEBgUOAszJLOXwAAiDAAIwAViJIJXc-nfaFU6XVerNTptbrrohUUj4HK0VabdiaYrXWTHe5xcPYkOFc6Z7I8IJjIQqvAAPyxBpO90inQtSyOJ1hpWkwMuNmJ5OpuceChUaJsvrJSl9PoLu7H2NLj6PIPw8ZIwQhE14XlEdl3gadlXsb13CDXBwSoMMIKoKC4NHYcENJcl41Ah8U3AZ8+gzV94AYP8ZDzUD3GMfUI2NGEagiBpqNomD8ToloGJ-cjMzXDctx3cBEkSRiPAaOj2zQiEeLHPMS1ErJxKoLj0Jouj-DoyImiM0CmnmNUQI1LVIB1VNmIAZRxQ8ZwA55UNLMMWTvdxSNTIF4CLNTQ3LawAHIWRCrtLN7azdXKaAAGFsF4HRMUgZovPAIodBKLAqlqeprHIMpKmqOpEGaeYspKajEAASR0FAMGQaxctKpAGAyUQqDQOTujghpK1MRrEErfF7JxIyGFy2AGla2pWnAOaamU4cGCGprTPyIA)

**Comment by brophdawg11** [comment, marked as answer]
`useMatches` does not return `RouteObject[]`, but since they share _some_ properties TS is allowing you to incorrectly define the type that way.  A "match" in this case is a mixtures of a matched route and the result of the routes `loader` function, so just remove `: RouteObject[]` and TS should be happy:

<img width="278" alt="Screenshot 2023-04-11 at 10 32 39 AM" src="https://user-images.githubusercontent.com/1609022/231197127-f4c63085-283d-4975-90b1-8a1b05a6baf9.png">

**Reply by brophdawg11** [reply]
It's hard to say without seeing your code. `useMatches` does not have an `infos` field on it if that's what you're trying to do.  Maybe you want to look for `match.data.infos`?
